kernel/generic/src/mm/as.c

```diff
@@ -1866,293 +1853,164 @@ size_t as_area_get_size(uintptr_t base)
-			uintptr_t left_pg = leaf->key[i - 1];
-			uintptr_t right_pg = leaf->key[i];
-			size_t left_cnt = (size_t) leaf->value[i - 1];
-			size_t right_cnt = (size_t) leaf->value[i];
+	return NULL;
+}
 
-			/*
-			 * The interval fits between left_pg and right_pg.
-			 */
+/** Get key function for used space ordered dictionary.
+ *
+ * The key is the virtual address of the first page
+ *
+ * @param odlink Ordered dictionary link (used_space_ival_t.lused_space)
+ * @return Pointer to virtual address of first page cast as @c void *.
+ */
+static void *used_space_getkey(odlink_t *odlink)
+{
+	used_space_ival_t *ival = odict_get_instance(odlink, used_space_ival_t,
+	    lused_space);
+	return (void *) &ival->page;
+}
 
-			if (overlaps(page, P2SZ(count), left_pg,
-			    P2SZ(left_cnt))) {
-				/*
-				 * The interval intersects with the left
-				 * interval.
-				 */
-				return false;
-			} else if (overlaps(page, P2SZ(count), right_pg,
-			    P2SZ(right_cnt))) {
-				/*
-				 * The interval intersects with the right
-				 * interval.
-				 */
-				return false;
-			} else if ((page == left_pg + P2SZ(left_cnt)) &&
-			    (page + P2SZ(count) == right_pg)) {
-				/*
-				 * The interval can be added by merging the two
-				 * already present intervals.
-				 */
-				leaf->value[i - 1] += count + right_cnt;
-				btree_remove(&area->used_space, right_pg, leaf);
-				goto success;
-			} else if (page == left_pg + P2SZ(left_cnt)) {
-				/*
-				 * The interval can be added by simply growing
-				 * the left interval.
-				 */
-				leaf->value[i - 1] += count;
-				goto success;
-			} else if (page + P2SZ(count) == right_pg) {
-				/*
-				 * The interval can be addded by simply moving
-				 * base of the right interval down and
-				 * increasing its size accordingly.
-				 */
-				leaf->value[i] += count;
-				leaf->key[i] = page;
-				goto success;
-			} else {
-				/*
-				 * The interval is between both neigbouring
-				 * intervals, but cannot be merged with any of
-				 * them.
-				 */
-				btree_insert(&area->used_space, page,
-				    (void *) count, leaf);
-				goto success;
-			}
-		}
-	}
+/** Compare function for used space ordered dictionary.
+ *
+ * @param a Pointer to virtual address of first page cast as @c void *
+ * @param b Pointer to virtual address of first page cast as @c void *
+ * @return Less than zero, zero, greater than zero if virtual address @a a
+ *         is less than, equal to, greater than virtual address b, respectively.
+ */
+static int used_space_cmp(void *a, void *b)
+{
+	uintptr_t va = *(uintptr_t *) a;
+	uintptr_t vb = *(uintptr_t *) b;
+
+	if (va < vb)
+		return -1;
+	else if (va == vb)
+		return 0;
+	else
+		return +1;
+}
 
-	panic("Inconsistency detected while adding %zu pages of used "
-	    "space at %p.", count, (void *) page);
+static void used_space_remove_ival(used_space_ival_t *ival)
+{
+	odict_remove(&ival->lused_space);
+	slab_free(used_space_ival_cache, ival);
+}
 
-success:
-	area->resident += count;
```

Comment:
I think the new code does not track area->resident anymore. This probably breaks statistics.

Reply:
OK, I think I need to move the used page count into used_space_t.